feat: multi chain asset list

vinnyhoward opened the pull request:
## **Description**

This PR introduces the Unified Asset List feature to MetaMask Mobile, providing users with a consolidated view of their assets across all supported blockchain networks. This enhancement improves the user experience by eliminating the need to switch between networks to view or manage assets, making asset management more intuitive and efficient.

We will followup with a PR to fix TS feedback we had!
(Not only TS issues but also [this](https://github.com/MetaMask/metamask-mobile/pull/12431#discussion_r1878052067 and [this](https://github.com/MetaMask/metamask-mobile/pull/12431#discussion_r1878048047))

## **Related issues**

Fixes: #12462 

## **Manual testing steps**

Build using `PORTFOLIO_VIEW` flag

```
PORTFOLIO_VIEW=true yarn watch
yarn start:ios 
yarn start:android
```

1. Go to the wallet page
2. Select all network on the network filter
3. Check the list of assets
4. Click on each asset with the network filter on "All Networks" and "Current Network"
5. Test send/swap flows with testnet networks to confirm everything still works
6. Importing all tokens should work when "All Networks" filter is on
7. Importing networks for a specific network should work when the "Current Network" filter is on
8. Aggregated balance should chance according to the network filter

## **Screenshots/Recordings**

| Before  | After |
|:---:|:---:|
|![before](https://github.com/user-attachments/assets/449bd3ef-1f69-4cf9-bb93-f5a28838e11b)|![after](https://github.com/user-attachments/assets/26209026-6863-4085-85ed-d541ca4fa720)|

### **Before**

<img src="https://github.com/user-attachments/assets/4f3aa5f5-920c-4ba6-8f13-0ce5bef4735f" width="350" alt="before_screenshot">

### **After**

<img src="https://github.com/user-attachments/assets/41e421b2-5b87-4933-a88e-159a45a15114" width="350" alt="after_screenshot">

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

salimtb committed:
feat: add multichain tokens autodetection feature

salimtb committed:
fix: fix hide token

salimtb committed:
fix: fix linter

vinnyhoward committed:
feat: cherry picking from multichain list experimental branch

salimtb committed:
fix: add unit test

salimtb committed:
fix: fix hide token case

salimtb committed:
fix: add unit tests

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-multichain-list

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-multichain-list

salimtb committed:
fix: clean up

vinnyhoward committed:
fix: fixed tests for feature flag off and all except one for feature flag on

vinnyhoward committed:
fix: linting issues

vinnyhoward committed:
fix: updating tests to increase coverage

gambinish commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -128,25 +185,80 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+            networkConfiguration.defaultRpcEndpointIndex
+          ]?.networkClientId;
+
+        NetworkController.setActiveNetwork(networkClientId as string).then(
+          () => {
+            setTimeout(() => {
+              navigation.navigate('Swaps', {
+                screen: 'SwapsAmountView',
+                params: {
+                  sourceToken:
+                    asset.address ?? swapsUtils.NATIVE_SWAPS_TOKEN_ADDRESS,
+                  sourcePage: 'MainView',
+                  chainId: asset.chainId,
+                },
+              });
+            }, 500);
+          },
+        );
```
Maybe we can `.catch` any errors here to stop execution, in the edge case that this fails.

vinnyhoward replied:
Yeah great idea, similar to what is being done for extension

vinnyhoward committed:
merged token detection branch and updated tests

gambinish commented on app/components/UI/Tokens/TokensBottomSheet/TokenFilterBottomSheet.tsx:
```diff
@@ -47,7 +58,9 @@ const TokenFilterBottomSheet = () => {
-  const isSelectedNetwork = Boolean(tokenNetworkFilter?.[chainId]);
+  const isCurrentNetwork = Boolean(
+    tokenNetworkFilter[chainId] && Object.keys(tokenNetworkFilter).length === 1,
+  );
```
It looks like the changes to this bottom sheet might be able to be separated into it's own PR?

One benefit that we saw in extension, was that breaking out a small, low risk piece like this, allowed us to also break out a lot of the fixture mocks needed for e2e. Made the final PR much smaller.

salimtb committed:
Merge branch 'main' into salim/multichain-detect-tokens-feat

salimtb committed:
fix: fix linter

salimtb committed:
Merge branch 'salim/multichain-detect-tokens-feat' into feat-multichain-list

salimtb committed:
fix: fix hide token

salimtb committed:
fix: increase test coverage

salimtb committed:
fix: increase test coverage

salimtb committed:
fix: add test for staking chain hook

salimtb committed:
fix: increase test coverage

salimtb committed:
fix: increase test coverage

salimtb committed:
fix: fix snapshot

salimtb committed:
fix: increase coverage

salimtb committed:
fix: increase test coverage

salimtb committed:
fix: increase test coverage

vinnyhoward commented on app/components/UI/Ramp/Views/BuildQuote/__snapshots__/BuildQuote.test.tsx.snap:
```diff
@@ -181,6 +181,7 @@ exports[`BuildQuote View Crypto Currency Data renders a special error page if cr
                 <View
                   style={
                     {
+                      "alignItems": "center",
```
Yeah I have two pending PRs and have some others in mind that can be broken up that will make this smaller and can be merged independently from PR
https://github.com/MetaMask/metamask-mobile/pull/12405
https://github.com/MetaMask/metamask-mobile/pull/12293

salimtb committed:
fix: fix block explorer

salimtb committed:
fix: fix bugs

salimtb committed:
fix: clean up

salimtb committed:
fix: clean up feature flag

salimtb committed:
fix: fix unit test

salimtb committed:
fix: fix PR comments

salimtb committed:
Merge branch 'main' into salim/multichain-detect-tokens-feat

salimtb committed:
fix: fix selected network

salimtb committed:
Merge branch 'main' into salim/multichain-detect-tokens-feat

salimtb committed:
Merge branch 'salim/multichain-detect-tokens-feat' into feat-multichain-list

vinnyhoward committed:
Merge branch 'salim/multichain-detect-tokens-feat' into feat-multichain-list

vinnyhoward committed:
fix: resolve memory leaks and mock function for feature flag causing snapshot to fail

salimtb committed:
fix: use allTokens

salimtb committed:
Merge branch 'main' into salim/multichain-detect-tokens-feat

salimtb committed:
fix: fix snapshot

salimtb committed:
fix: fix snapshot

salimtb committed:
Merge branch 'salim/multichain-detect-tokens-feat' into feat-multichain-list

salimtb committed:
fix: fix after conflicts

sahar-fehri committed:
feat: aggregated balance cross chains for portfolio view

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-multichain-list

vinnyhoward committed:
fix: lint

vinnyhoward committed:
fix: increase test coverage and fix asset details page not showing balance on some tokens

vinnyhoward committed:
fix: e2e with feature flag off is fixed and added more test coverage

vinnyhoward committed:
fix: fixed tests

sahar-fehri committed:
Merge branch 'feat/cross-chain-aggregated-balance-feature' into feat-multichain-list

sahar-fehri committed:
feat: fix lint

sahar-fehri committed:
fix: fix native token balance display when zero

salimtb committed:
fix: filter system files

jonybur commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -94,6 +126,8 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
 
   useEffect(() => {
+    // TODO: Replace "any" with type
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Not blocker: We should get this TODO either fixed or linked to a ticket

vinnyhoward replied:
It was updated with a type 2 weeks ago but the comment wasn't removed. I'll remove now

jonybur commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -128,25 +185,80 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+        });
+      }
+    } else {
+      navigation.navigate('Swaps', {
```
Seems like we are repeating this block:
```
        navigation.navigate('Swaps', {
          screen: 'SwapsAmountView',
          params: {
            sourceToken: asset.address ?? swapsUtils.NATIVE_SWAPS_TOKEN_ADDRESS,
            sourcePage: 'MainView',
            chainId: asset.chainId,
          },
        });
```
There's probably a way to alter the logic so that we dont have repeated blocks like that

jonybur commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -209,57 +321,86 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
-      ? balanceFiat
-      : `${balance} ${asset.symbol}`;
+    mainBalance = `${balance} ${asset.symbol}`;
+    secondaryBalance = asset.balanceFiat;
```
Same here, seems like our code is repeating blocks of logic across several components

vinnyhoward replied:
This is because of some feature flag logic. This will be simplified once the feature flag is removed

jonybur commented on app/components/UI/Stake/hooks/useStakingChain.ts:
```diff
@@ -13,4 +14,15 @@ const useStakingChain = () => {
 
+export const useStakingChainByChainId = (chainId: Hex) => {
+  // Updated to accept chainId as a parameter
+  // const chainId = useSelector(selectChainId); // Removed the useSelector hook
```
Remove this line?

jonybur commented on app/selectors/multichain.ts:
```diff
@@ -1,4 +1,289 @@
+          }).format(tokenFiatAmount);
+        }
+        const name = isETH ? `Ethereum` : nativeCurrency;
+        // TODO: Check if this is correct for native tokens
```
Is this TODO still necessary?

vinnyhoward replied:
No its no longer needed. I'll remove

jonybur commented on app/selectors/multichain.ts:
```diff
@@ -1,4 +1,289 @@
+        tokensByChain[currentChainId] = userTokens.map((token) => {
+          const tokenAddress = token.address as Hex;
+          const tokenExchangeRatePriceByTokenAddress =
+            // TODO: Some exchange rates for some tokens don't exist? Is this expected?
```
TODO?

vinnyhoward replied:
Old comment and no longer an issue. I'll remove TODO

salimtb committed:
fix: use swap vertical icon

jonybur commented on app/selectors/tokensController.ts:
```diff
@@ -16,10 +13,23 @@ import {
 const selectTokensControllerState = (state: RootState) =>
   state?.engine?.backgroundState?.TokensController;
 
+// TODO: Double check this make sure this allTokens change plays nice with original view
```
TODO here, could be covered with a test

vinnyhoward replied:
Tests have recently been udpated

sahar-fehri committed:
fix: fix swaps button display

sahar-fehri committed:
fix: feedback

vinnyhoward committed:
fix: added more test coverage, cleaned up todos, reverted unnecessary changes

vinnyhoward committed:
Merge branch 'feat/cross-chain-aggregated-balance-feature' of github.com:MetaMask/metamask-mobile into feat-multichain-list

vinnyhoward committed:
test: increase test coverage for multichain file and fixed incorrect import

sahar-fehri committed:
fix: fix error when showConversion is not enabled for testnet

salimtb committed:
fix: fix send flow testnet

vinnyhoward committed:
fix(12550): adding popular network no longer switches network filter (#12578)

## **Description**

Fixed an issue where adding a popular network switches the network
filter. For example if the user is on "All Networks" and they add a
popular network such as Avalanche, the network switcher reverts to that
network, instead it should stay at "All Networks"

## **Related issues**

Fixes:
[#12550](https://github.com/MetaMask/metamask-mobile/issues/12550)

## **Manual testing steps**

1. Click on "All Networks" filter
2. Click on the network tab, then add a popular network such as
Avalanche
3. Make sure the network filter stays at "All Networks"

## **Screenshots/Recordings**

NA

### **Before**

See video of before in
[issue](https://github.com/MetaMask/metamask-mobile/issues/12550)

### **After**


https://github.com/user-attachments/assets/49577d3d-d329-4cfd-8341-dee70c33d754

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-multichain-list

vinnyhoward committed:
fix: removed dup import

vinnyhoward committed:
fix: add missing symbol in main balance

vinnyhoward committed:
fix: optimized multichain list so that it re-renders less

vinnyhoward committed:
Merge branch 'main' into feat-multichain-list

vinnyhoward committed:
test: update multichain test

vinnyhoward committed:
fix: lint

sahar-fehri committed:
fix: fix balance

salimtb committed:
Merge branch 'main' into feat-multichain-list

sahar-fehri committed:
fix: fix showing fiat balance on testnet when conversion rate is off

sahar-fehri committed:
fix: fix merge conflicts

vinnyhoward committed:
fix: refactored multichain list to be faster and simpler

vinnyhoward committed:
Merge branch 'feat-multichain-list' of github.com:MetaMask/metamask-mobile into feat-multichain-list

salimtb committed:
fix: fix balance

salimtb committed:
fix: fix linter

salimtb committed:
fix: fix linter

salimtb committed:
fix: fix native/fiat toggle

salimtb committed:
fix: clean up

salimtb committed:
fix: clean up and reduce review scope

salimtb committed:
fix: fix balance + detection

salimtb committed:
fix: fix unit tests

vinnyhoward committed:
Merge branch 'main' into feat-multichain-list

vinnyhoward committed:
fix: added fallbacks if there is missing markets, token balances

salimtb committed:
fix: remove conf file from diff

sahar-fehri committed:
fix: fix conversions for testnets

salimtb committed:
Merge branch 'main' into feat-multichain-list

salimtb committed:
fix: fix after conflicts

salimtb committed:
fix: fix after conflicts

vinnyhoward committed:
fix: native balance address

vinnyhoward committed:
fix: token balances render again for non-native tokens in asset overview component

vinnyhoward committed:
fix: bnb and polygon returning network names and not token names

vinnyhoward committed:
fix: fixed issue where POL and BNB wasn't showing balance in asset overview component

vinnyhoward committed:
fix: fixed balance issues relating to non-ethereum native tokens

vinnyhoward committed:
fix: update snap

Matt561 commented on app/components/UI/Stake/components/StakingBalance/StakingBalance.test.tsx:
```diff
@@ -122,7 +123,7 @@ describe('StakingBalance', () => {
   });
 
-  it('render matches snapshot', () => {
+  (isPortfolioViewEnabled() ? it.skip : it)('render matches snapshot', () => {
```
Is this temporary while another feature is in progress? Ideally we wouldn't want to conditionally run tests.

vinnyhoward replied:
Noted! I'll fix this now

vinnyhoward replied:
Just updated it. It now doesn't conditionally  skip but runs a separate test specifically for the feature flag snapshot

Matt561 reviewed (commented):
I left a comment related to running a staking snapshot test conditionally. Otherwise, staking-owned files look good to me!

sahar-fehri committed:
Merge branch 'main' into feat-multichain-list

vinnyhoward committed:
fix: tests no longer skip due to feature flags but run conditionally

vinnyhoward committed:
fix: update snapshot

salimtb committed:
fix: reduce code owners

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -79,12 +98,35 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+  const selectedTicker = useSelector((state: RootState) => selectTicker(state));
+
+  const nativeCurrency = useSelector((state: RootState) =>
+    selectNativeCurrencyByChainId(state, asset.chainId as Hex),
```
Is there the possibility to fix the type instead of cast here?

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -79,12 +98,35 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+  let currentAddress: Hex;
+
+  if (isPortfolioViewEnabled()) {
+    currentAddress = asset.address as Hex;
```
Same here

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -79,12 +98,35 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+    currentAddress = asset.address as Hex;
+  } else {
+    currentAddress = asset.isETH
+      ? getNativeTokenAddress(chainId as Hex)
+      : (asset.address as Hex);
```
Same here

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -119,7 +161,41 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+            networkConfiguration.defaultRpcEndpointIndex
+          ]?.networkClientId;
+
+        await NetworkController.setActiveNetwork(networkClientId as string);
```
For thr networkCkientId as well

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -230,9 +346,22 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+    const multiChainTokenBalanceHex =
+      itemAddress &&
+      multiChainTokenBalance?.[selectedInternalAccountAddress as Hex]?.[
+        chainId as Hex
```
chainId should be hex already no? I wonder why typescript complains about it

sahar-fehri replied:
Yes i think we could remove the cast on this one;
Thank you @tommasini  for your feedback 🙏  We have plans to do a follow up PR to refactor; and  we will address your ts comments on a followup PR.

Unless its a major thing that we should address; i think we can followup on this PR 🙏

tommasini replied:
Sounds great!

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -243,23 +372,37 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
-      ? balanceFiat
-      : `${balance} ${asset.symbol}`;
+    mainBalance = `${balance} ${asset.ticker}`;
+    secondaryBalance = exchangeRate ? asset.balanceFiat : '';
```
This fallbackk changed, flagging this line to understand what happens if secondaryBalance is an empty string

sahar-fehri replied:
Hi @tommasini if this is empty then it wont be displayed in the UI; I think the logic has slightly changed this is why the fallback also have changed

tommasini replied:
Thank you for the explanation!

tommasini commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -128,25 +204,58 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
+      }
+    } else {
+      handleSwapNavigation();
+      trackEvent(
```
Does it make sense to move the trackEvent to `handleSwapNavigation` since we want to track the navigation from the TokenDetails to swap I think it's still accurate if we change it to there

tommasini commented on app/components/UI/AssetOverview/TokenDetails/TokenDetails.tsx:
```diff
@@ -46,20 +54,36 @@ interface TokenDetailsProps {
-  const tokenExchangeRates = useSelector(selectContractExchangeRates);
-  const conversionRate = useSelector(selectConversionRate);
+  const tokenExchangeRatesByChainId = useSelector((state: RootState) =>
+    selectTokenMarketDataByChainId(state, asset.chainId as Hex),
```
Flagging this cast

tommasini commented on app/components/UI/AssetOverview/TokenDetails/TokenDetails.tsx:
```diff
@@ -46,20 +54,36 @@ interface TokenDetailsProps {
+    selectTokenMarketDataByChainId(state, asset.chainId as Hex),
+  );
+  const nativeCurrency = useSelector((state: RootState) =>
+    selectNativeCurrencyByChainId(state, asset.chainId as Hex),
```
flagging this cast

tommasini commented on app/components/UI/AssetOverview/TokenDetails/TokenDetails.tsx:
```diff
@@ -46,20 +54,36 @@ interface TokenDetailsProps {
+  } else if (tokenContractAddress) {
     tokenMetadata = tokenList?.[tokenContractAddress.toLowerCase()];
-    marketData = tokenExchangeRates?.[tokenContractAddress];
+    marketData = tokenExchangeRates?.[tokenContractAddress as Hex];
```
This cast is new, most likely because the logic behind the portfolio view could be better typed or we have a new fallback to handle?

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
-  const chainId = useSelector(selectChainId);
+  const selectedChainId = useSelector(selectChainId);
+  const chainId = isPortfolioViewEnabled()
+    ? (asset.chainId as Hex)
```
Flagging this cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+    ? multiChainMarketData?.[chainId as Hex]
+    : singleTokenExchangeRates;
+  const tokenBalances = isPortfolioViewEnabled()
+    ? multiChainTokenBalance?.[selectedInternalAccountAddress as Hex]?.[
```
flagging this cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+    : singleTokenExchangeRates;
+  const tokenBalances = isPortfolioViewEnabled()
+    ? multiChainTokenBalance?.[selectedInternalAccountAddress as Hex]?.[
+        chainId as Hex
```
Flagging this cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+      ]
+    : selectedChainTokenBalance;
+  const nativeCurrency =
+    networkConfigurations?.[chainId as Hex]?.nativeCurrency;
```
flagging this cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
-      tokenBalances,
-      conversionRate,
-      currentCurrency,
+      exchangeRates || {},
```
It seems that any of this fallbacks is harmless, although, If we are doing a follow up PR to fix typing, can we please fix this anys: https://github.com/MetaMask/metamask-mobile/blob/b2f9eecdd24008b0338a0f3fb5bb6a134dec2040/app/components/UI/Tokens/util/deriveBalanceFromAssetMarketDetails.ts#L12

sahar-fehri replied:
I see that the "any" in the `deriveBalanceFromAssetMarketDetails` fct were not introduced specifically in this PR but i agree that we need to address it; can we follow up on this one too 🙏

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+      exchangeRates || {},
+      tokenBalances || {},
+      conversionRate || 0,
+      currentCurrency || '',
```
Just confirming that we want to show no current currency if this fallsback to an empty string
https://github.com/MetaMask/metamask-mobile/blob/8229dbcc07de3506ab3e0d1082502c2618ea1cc9/app/util/number/index.js#L525

If so we should fix it on `deriveBalanceFromAssetMarketDetails ` instead of where it's called right?
Since it's used in [Tokens](https://github.com/MetaMask/metamask-mobile/blob/eefb4280e5e4c2834b3383e2fbaae2e6e9a158c5/app/components/UI/Tokens/index.tsx#L131) and [TokenListItem](https://github.com/MetaMask/metamask-mobile/blob/b2f9eecdd24008b0338a0f3fb5bb6a134dec2040/app/components/UI/Tokens/TokenList/TokenListItem/index.tsx#L93) this will help have parity while the function itself handles the fallbacks.

salimtb replied:
If I understand correctly, you're suggesting that we set default values directly in the `deriveBalanceFromAssetMarketDetails` function instead of implementing the fallback in this way. We prefer to make these changes in a separate PR when the feature flag is set to `true`. We don’t believe this approach will cause any issues if we leave it as is for now.

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+
+  if (isPortfolioViewEnabled()) {
+    const tokenPercentageChange = asset.address
+      ? multiChainMarketData?.[chainId as Hex]?.[asset.address as Hex]
```
flag cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+      : 0;
+
+    pricePercentChange1d = asset.isNative
+      ? multiChainMarketData?.[chainId as Hex]?.[zeroAddress() as Hex]
```
flag cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+      : tokenPercentageChange;
+  } else {
+    pricePercentChange1d = itemAddress
+      ? exchangeRates?.[itemAddress as Hex]?.pricePercentChange1d
```
flag cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -67,75 +82,117 @@ export const TokenListItem = ({
+  } else {
+    pricePercentChange1d = itemAddress
+      ? exchangeRates?.[itemAddress as Hex]?.pricePercentChange1d
+      : exchangeRates?.[zeroAddress() as Hex]?.pricePercentChange1d;
```
flag cast

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -154,38 +211,105 @@ export const TokenListItem = ({
+      if (isTestNet(currentChainId))
+        return getTestNetImageByChainId(currentChainId);
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const defaultNetwork = getDefaultNetworkByChainId(currentChainId) as any;
```
We should never use any

sahar-fehri replied:
I think any was used here because `getDefaultNetworkByChainId` was not typed; 
Exp of another usage on main: https://github.com/MetaMask/metamask-mobile/blob/9d63e3407607e32e7fadc2b33c2d776193d930a3/app/util/networks/index.js#L405

tommasini commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -100,22 +109,118 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
+      // MultiChain implementation
+      const allTokens = Object.values(
+        selectedAccountTokensChains,
+      ).flat() as TokenI[];
```
Flagging cast

tommasini commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -130,10 +235,10 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
-                tokenBalances,
-                conversionRate,
-                currentCurrency,
+                tokenExchangeRates || {},
```
Check here: https://github.com/MetaMask/metamask-mobile/pull/12431/files#r1877933307

salimtb replied:
answer on the flagged message

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -154,38 +211,105 @@ export const TokenListItem = ({
+    if (isPortfolioViewEnabled() && asset.isNative) {
+      return (
+        <NetworkAssetLogo
+          chainId={chainId as Hex}
```
Flag cast

sahar-fehri committed:
fix: fix getDefaultNetworkByChainId any

tommasini commented on app/components/UI/Tokens/util/filterAssets.ts:
```diff
@@ -0,0 +1,62 @@
+  range: (value: number, opts: Record<string, number>) => boolean;
+}
+
+const filterCallbacks: FilterCallbacksT = {
```
Can we add a TSDoc explanation here?

tommasini commented on app/components/UI/Tokens/util/filterAssets.ts:
```diff
@@ -0,0 +1,62 @@
+  return get(obj, keyPath);
+}
+
+export function filterAssets<T>(assets: T[], criteria: FilterCriteria[]): T[] {
```
And here it would be good to have a TSDoc as well

tommasini commented on app/components/Views/AssetOptions/AssetOptions.tsx:
```diff
@@ -146,7 +181,18 @@ const AssetOptions = (props: Props) => {
+
+              const networkClientId =
+                NetworkController.findNetworkClientIdByChainId(
+                  chainIdToUse as Hex,
```
Flagging cast

tommasini commented on app/components/Views/Wallet/index.tsx:
```diff
@@ -390,23 +391,31 @@ const Wallet = ({
+          const networkClientId =
+            rpcEndpoints[defaultRpcEndpointIndex]?.networkClientId;
+          if (networkClientId) {
+            return AccountTrackerController.refresh(networkClientId);
```
Curious question, is there any plan to optimise this[ refresh function](https://github.com/MetaMask/core/blob/f5ad522451002eb2b56fdac78d2b11ad26ac471a/packages/assets-controllers/src/AccountTrackerController.ts#L356), since we use mutex, I believe this will only be called once, I wonder if we could architect a `refreshAll` function to accept multiple networkClientIds, I believe that way we could be able to improve the performance of this! Food for thoughts!

tommasini replied:
We are calling all this promises at once, which makes me doubt if the mutex under the hood really works for all of them or just for the first or just ignore some of the `networkClietId` that receives while the other is still running

sahar-fehri replied:
We updated this code and removed the promise.all; thnx for flagging this 🙏

salimtb commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -154,38 +211,105 @@ export const TokenListItem = ({
+        <NetworkAssetLogo
+          chainId={chainId as Hex}
+          style={styles.ethLogo}
+          ticker={asset.ticker || ''}
```
this is a small nit , we will address this on a separated follow up PR

tommasini commented on app/reducers/swaps/index.js:
```diff
@@ -190,6 +194,39 @@ const swapsControllerAndUserTokens = createSelector(
   },
 );
 
+const swapsControllerAndUserTokensMultichain = createSelector(
```
Could we add TSDdoc and unit tests for this selector?

tommasini commented on app/selectors/currencyRateController.ts:
```diff
@@ -27,6 +27,12 @@ export const selectConversionRate = createSelector(
+export const selectCurrencyRates = createSelector(
+  selectCurrencyRateControllerState,
+  (currencyRateControllerState: CurrencyRateState) =>
+    currencyRateControllerState?.currencyRates,
```
What happens if this is undefined? It would be good for we to start handle this scenarios directly on the selector to type the way the consumers components use the data

tommasini commented on locales/languages/en.json:
```diff
@@ -470,6 +470,7 @@
     "learn_more": "Learn more",
     "add_collectibles": "Import NFTs",
     "no_transactions": "You have no transactions!",
+    "switch_network_to_view_transactions": "Please switch network to view transactions",
```
We can't merge this and cut it directly to the release, will not include the translations for this copy

sahar-fehri replied:
I think we can ship this separately to no block this one 🙏

salimtb committed:
fix: add TS doc

tommasini commented on patches/@metamask+assets-controllers+45.1.1.patch:
```diff
@@ -161,3 +161,109 @@ index a34725f..12487d6 100644
++      __classPrivateFieldGet(this, _TokenDetectionController_instances, "m", _TokenDetectionController_stopPolling).call(this);
++  });
++  this.messagingSystem.subscribe('TokenListController:stateChange',
++  // TODO: Either fix this lint violation or explain why it's necessary to ignore.
```
Flagging a todo

tommasini commented on patches/@metamask+assets-controllers+45.1.1.patch:
```diff
@@ -161,3 +161,109 @@ index a34725f..12487d6 100644
+-            });
+-        }
+-    });
++      const { networkConfigurationsByChainId } = this.messagingSystem.call('NetworkController:getState');
```
Can we have a comment explaining the change patched and if possible the PR or issue where the patch will be addressed?

salimtb replied:
related to this: https://github.com/MetaMask/core/pull/4967

salimtb committed:
fix: address refresh mutex

salimtb committed:
fix: clean up

sahar-fehri committed:
fix: rm new locales

sahar-fehri commented:
Thanks @tommasini again 🙏  we will address your TS feedback on a followup PR 🙏

sahar-fehri committed:
fix: fix

sahar-fehri committed:
Merge branch 'main' into feat-multichain-list

tommasini reviewed (approved):
LGTM! can we add the follow up steps at the PR description?

Matt561 reviewed (approved):
Approved on behalf of the staking team!

salimtb committed:
Merge branch 'main' into feat-multichain-list
